proofs/alf/cvc5/theories/BitVectors.smt3

```diff
@@ -5,32 +5,49 @@
 
 ; Returns the (integer value) bitwidth for a given bit-vector type.
 (define $bv_bitwidth ((T Type))
-  (alf.match ((n Int)) T (((BitVec n) n))))
+  (alf.match ((n Int)) T (((BitVec n) n)))
+)
+
+; bvsize
+(declare-const @bvsize (-> (! Int :var m :implicit) (BitVec m) Int))
+
+; evaluator for terms that may impact the type of nodes
+(program $eval_bv_sym ((S Type) (T Type) (f (-> T S)) (x S) (ys S :list) (z S) (y T) (m Int) (xb (BitVec m)))
+    (S) S
+    (
+    (($eval_bv_sym (+ x ys))     (arith_eval_add ($eval_bv_sym x) ($eval_bv_sym ys)))
+    (($eval_bv_sym (- x z))      (arith_eval_sub ($eval_bv_sym x) ($eval_bv_sym z)))
+    (($eval_bv_sym (@bvsize xb)) ($bv_bitwidth (alf.typeof xb)))
+    (($eval_bv_sym (int.log2 x)) (arith_eval_int_log_2 ($eval_bv_sym x)))
+    (($eval_bv_sym (mod x z))    (alf.zmod ($eval_bv_sym x) ($eval_bv_sym z)))
+    (($eval_bv_sym x)            x)
+    )
+)
 
 ; used as the nil terminator for concat
-(define @bvempty () (alf.to_bin 0 0))
+(define @bv_empty () (alf.to_bin 0 0))
 
 (declare-const concat (->
   (! Int :var n :implicit)
   (! Int :var m :implicit)
   (BitVec n)
   (BitVec m)
-  (BitVec (alf.add n m))) :right-assoc-nil @bvempty
+  (BitVec (alf.add n m))) :right-assoc-nil @bv_empty
 )
 
 (declare-const extract (->
   (! Int :var n :implicit)
   (! Int :var h)
-  (! Int :var l :requires ((alf.is_neg l) false))
+  (! Int :var l)
```

Comment:
Why?

Reply:
This is because `l` is no longer required to be a positive integer.

Reply:
I could in theory make this test ensure that `l` does not evaluate to a negative integer. although this is a bit of overkill.